streams/src/main/java/org/apache/kafka/streams/processor/internals/InternalTopicManager.java

```diff
@@ -63,87 +78,122 @@ public InternalTopicManager(final StreamsKafkaClient streamsKafkaClient,
+                    final String topicName = createTopicResult.getKey();
+                    verifyIfTopicExistsWithCorrectNumberOfPartitions(topicName, topics.get(topicName));
+                } catch (final InterruptedException fatalException) {
+                    final String error = "Could not create internal topics.";
+                    log.error(error, fatalException);
+                    throw new StreamsException(error, fatalException);
+                }
+            }
+        }
+    }
+
+    private void verifyIfTopicExistsWithCorrectNumberOfPartitions(final String topic,
+                                                                  final InternalTopicConfig expectedNumberOfPartitions) {
+        final Map<String, Integer> existingTopicPartitions = getNumPartitions(Collections.singleton(topic));
+        if (validateTopicPartitions(Collections.singleton(expectedNumberOfPartitions), existingTopicPartitions).size() != 0) {
+            final String errorMsg = String.format("Could neither create topic %s nor receive any metadata about it.", topic);
+            log.error(errorMsg);
+            throw new StreamsException(errorMsg);
         }
-        throw new StreamsException("Could not create internal topics.");
     }
 
     /**
      * Get the number of partitions for the given topics
      */
     public Map<String, Integer> getNumPartitions(final Set<String> topics) {
-        for (int i = 0; i < MAX_TOPIC_READY_TRY; i++) {
-            try {
-                final MetadataResponse metadata = streamsKafkaClient.fetchMetadata();
-                final Map<String, Integer> existingTopicPartitions = fetchExistingPartitionCountByTopic(metadata);
-                existingTopicPartitions.keySet().retainAll(topics);
-
-                return existingTopicPartitions;
-            } catch (StreamsException ex) {
-                log.warn("Could not get number of partitions: {} Retry #{}", ex.getMessage(), i);
+        return getNumPartitions(topics, false);
+    }
+
+    private Map<String, Integer> getNumPartitions(final Set<String> topics,
+                                                  final boolean bestEffort) {
+        try {
+            final DescribeTopicsResult describeTopicsResult = adminClient.describeTopics(topics);
+            final Map<String, KafkaFuture<TopicDescription>> futures = describeTopicsResult.values();
+
+            final Map<String, Integer> existingTopicPartitions = new HashMap<>();
+            for (final Map.Entry<String, KafkaFuture<TopicDescription>> topicFuture : futures.entrySet()) {
+                try {
+                    final TopicDescription topicDescription = topicFuture.getValue().get();
+                    existingTopicPartitions.put(
+                        topicFuture.getKey(),
+                        topicDescription.partitions().size());
+                } catch (final ExecutionException topicDoesNotExist) {
```

Comment:
Ditto here for different exception types.

Reply:
For this response we could get
```
    /**
     * Possible error codes:
     *
     * REQUEST_TIMED_OUT(7)
     * INVALID_TOPIC_EXCEPTION(17)
     * TOPIC_AUTHORIZATION_FAILED(29)
     * NOT_CONTROLLER(41)
     */
```

All seem to be fatal. Comments?

Reply:
Here we get the following:
```
     * UnknownTopic (3)
     * LeaderNotAvailable (5)
     * InvalidTopic (17)
     * TopicAuthorizationFailed (29)
```

and 

```
     * LeaderNotAvailable (5)
     * ReplicaNotAvailable (9)
```

and

```
     * STALE_CONTROLLER_EPOCH (11)
```

All seem to be fatal, considering internal `AdminClient` retries.

Reply:
Make sense.